Authentication Interceptor 와 WebConfig 구현

pjkfckr opened the pull request:
Interceptor에서 Request Path 와 Request Method 로 filtering 구현
Interceptor에서 header의 Authorization 을 가져와 토큰을 검사하는 doAuthentication 메소드 구현
WebConfig 클래스에서 AuthenticationInterceptor를 등록

#### 알아가는 점
현재 사용중인 프레임워크에서 Filter 와 Interceptor를 구현하여 중복된 코드 작성을 없애기 위해 사용하고 있었는데,
이를 Spring 에서 마주쳐서 반가웠습니다.
제가 원하는 방향은 `annotation` 으로 `Authorization`을 만들어 해당 어노테이션이 있는 Method만 AuthenticationInterceptor를 통하게끔 변경하고 싶었는데, 시도 중 우선은 강의를 따라서 해보는 것 또한 큰 의미가 있을것 같아 강의대로 먼저 PR를 날렸습니다.

pjkfckr committed:
Authentication Interceptor 추가 와 WebConfig 구현

Interceptor에서 Request Path 와 Request Method 로 filtering 구현
Interceptor에서 header의 Authorization 을 가져와 토큰을 검사하는 doAuthentication 메소드 구현
WebConfig 클래스에서 AuthenticationInterceptor를 등록

johngrib commented on app/src/main/java/com/codesoom/assignment/interceptors/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,68 @@
+        String path = request.getRequestURI();
+        String method = request.getMethod();
+        if (path.startsWith("/products") && (
+                method.equals("GET") || method.equals("OPTIONS"))
```
`method`가 `null`인 경우에 여기에서 NPE가 터질 위험이 있습니다. 이렇게 수정해 봅시다.
```suggestion
                "GET".equals(method) || "OPTIONS".equals(method))
```

pjkfckr replied:
NPE에 대한 생각을 전혀 하지 못했는데, 앞으로도 다시한번 생각해보겠습니다.

johngrib replied:
ㅎㅎ 이 기법은 자바에서 흔한 패턴이니 꼭 기억해두세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/interceptors/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    /**
+     * 유효한 토큰으로 요청을 했는지 확인
```
boolean 메소드는 이런 방식으로 JavaDoc을 작성해 보세요.

```suggestion
     * 유효한 토큰으로 요청했다면 true, 그렇지 않다면 false를 리턴합니다.
```

pjkfckr replied:
boolean을 반환하는 메소드는 어떤상황에서 true와 false가 발생하는지를 작성하는 것이 좋은 작성방법 이네요!
만약 Optional 을 반환하는 경우에는 `Optional.empty` 와 `Optional.of` 인 두가지의 경우를 서술하는 것이 좋을까요?

johngrib replied:
Optional이라면 그렇게까지 하지 않아도 괜찮아요. 하지만 설명이 필요한 경우도 있으니 양쪽을 잘 저울질해봐야겠죠.

johngrib commented:
좋습니다! 다만 PR 본문이 좀 더 매끄러운 글이 되어도 좋을 것 같아요. 다음은 @SUBIN606 님의 PR에 남긴 코멘트입니다. 한번 읽어보시고 어떻게 PR 본문을 적을지 궁리해 보세요.

https://github.com/CodeSoom/spring-week7-assignment-1/pull/52#pullrequestreview-964568446

pjkfckr committed:
AuthenticationInterceptor 수정

- filterWithPathAndMethod 메소드에서 NPE를 고려하여 method 가 아닌 ["GET", "OPTIONS", "POST"] 의 equals 비교로 변경
- boolean 메소드의 JavaDoc 작성시 true와 false를 반환하는 조건에 대한 설명추가

pjkfckr commented:
#### Interceptor에서 검증이 필요없는 Request의 경우 검증하지않도록 구현

- Interceptor에서 검증이 필요없는 Path와 Method 를 filtering하는 메소드 추가 하였습니다.

- 로그인, 상품 조회, 회원가입을 하는 경우에는 검증이 필요없기때문입니다.

- Path 와 Method로 구분하여 로그인, 상품조회, 회원가입으로 요청한 Request의 경우 검증하지않도록  `filterWithPathAndMethod` 메소드를 구현했습니다.

#### Interceptor에서 JWT 를 검사하도록 구현
- 검증이 필요한 API의 경우 해당 사용자가 권한이 있는지 확인하도록 구현하였습니다.

- 모든 사용자가 상품을 수정, 삭제 하게되면 일반 사용자가 모든 상품을 수정, 삭제를 할수있기때문입니다.

- `Request` 에서 `Header` 에있는 `Authorization을 가져와 해당 JWT가 유효한지 확인 한 후 유효하다면 true를 반환, 유효하지 않다면 false를 반환하는 `doAuthentication` 메소드를 구현했습니다.

#### `Authentication Interceptor` 를 등록
- 위에서 구현한 `Interceptor`를 등록하여 사용할수 있도록 구현

- 구현한 `Interceptor` 를 등록하지 않는다면 Spring 에서 해당 `Interceptor`가 있는지도 모르기때문입니다.

- `InterceptorRegistry` class에 `addInterceptor` 메소드를 사용하여 구현한 `Interceptor`를 등록했습니다.

pjkfckr commented:
위와 같이 작성을 해보았습니다. `what` `why` `how` 에 최대한 맞추어서 작성해보았는데, PR 본문을 한번 적고 난 후 리뷰 후에 다시 commit 한 내용들은 또 이렇게 본문을 작성하여 올리는 것인지, 아니면 추가 또는 리뷰와 다른 내용을 수정한 경우에 본문을 작성하는 것인지 궁금합니다.

johngrib commented:
> 위와 같이 작성을 해보았습니다. `what` `why` `how` 에 최대한 맞추어서 작성해보았는데, PR 본문을 한번 적고 난 후 리뷰 후에 다시 commit 한 내용들은 또 이렇게 본문을 작성하여 올리는 것인지, 아니면 추가 또는 리뷰와 다른 내용을 수정한 경우에 본문을 작성하는 것인지 궁금합니다.

이렇게 아래쪽이 이어서 작성해주시면 됩니다. 다음 PR을 기대해볼게요.

johngrib reviewed (approved):
요청한 내용대로 잘 변경해주셨습니다. 수고하셨습니다.
